src/main/java/pink/zak/minestom/towerdefence/ui/spawner/TroopSpawnerUI.java

```diff
@@ -0,0 +1,180 @@
+
+        EnemyMob enemyMob = optionalEnemyMob.get();
+        int levelInt = this.gameUser.getMobLevel(enemyMob);
+        EnemyMobLevel level = enemyMob.getLevel(levelInt);
+
+        if (this.mode == Mode.SEND && level != null) attemptToSendMob(enemyMob);
+        if (this.mode == Mode.UPGRADE) setTab(new TroopUpgradeTab(this, this.gameUser, enemyMob));
+    }
+
+    private void attemptToSendMob(@NotNull EnemyMob mob) {
+        GameUser user = this.gameUser;
+
+        EnemyMobLevel level = mob.getLevel(user.getMobLevel(mob));
+        if (level == null) return; // the player attempted to send a mob that they don't own
+
+        // check if the player can afford to send the mob
+        if (user.getCoins() < level.getSendCost()) return;
+        if (!user.getQueue().canQueue(mob)) return;
+
+        // charge player for sending the mob
+        user.updateCoins(current -> current - level.getSendCost());
+
+        // add the mob to the queue
+        user.getQueue().queue(new QueuedEnemyMob(mob, level));
+        this.updateQueue(user.getQueue().getQueuedMobs());
+    }
+
+    private void enterUpgradeMode() {
+        this.mode = Mode.UPGRADE;
+        this.setTitle(UPGRADE_TITLE);
+        this.setItemStack(31, ItemStack.AIR);
+        this.setItemStack(35, ItemStack.AIR);
+    }
+
+    private void setTab(@NotNull TroopUpgradeTab tab) {
+        this.tab = tab;
+        updateSubInventory();
+    }
+
+    void updateSubInventory() {
+        if (this.tab == null) return;
+        for (int i = 9; i < 27; i++) setItemStack(i, this.tab.getItemStack(i - 9));
+    }
+
+    private @NotNull ItemStack createQueueItem(@NotNull Queue<QueuedEnemyMob> queue) {
+        List<ItemStack> items = new ArrayList<>();
+
+        EnemyMob currentlyTrackedMob = null;
+        int count = 0; // count of the current iterated mob
+        for (QueuedEnemyMob queuedMob : queue) {
+            EnemyMob mob = queuedMob.mob();
+            if (currentlyTrackedMob != mob) {
+                if (currentlyTrackedMob != null) {
+                    items.add(currentlyTrackedMob.getLevel(this.gameUser.getMobLevels().get(currentlyTrackedMob)).createPreviewItem().withAmount(count));
+                    count = 0;
+                }
+
+                // we +1 at the end, so it gets to 150 :)
+                // also this is an arbitrary limit, not Minecraft's.
+                if (items.size() == 149) break;
+
+                currentlyTrackedMob = mob;
+            }
+            count++;
+        }
+
+        if (currentlyTrackedMob != null) items.add(currentlyTrackedMob.getLevel(this.gameUser.getMobLevels().get(currentlyTrackedMob)).createPreviewItem().withAmount(count));
+
+        return BASE_QUEUE_ITEM.withMeta(BundleMeta.class, meta -> meta.items(items));
+    }
+
+    public void updateQueue(@NotNull Queue<QueuedEnemyMob> queue) {
+        setItemStack(35, createQueueItem(queue));
+    }
```

Comment:
Since queues are player-specific, this update can be called inside the inventory as the queue will only be updated after interacting with it.

Reply:
This cannot be done as the queue updates itself, maybe an event can be used instead?